File: 2025-9322-team04_finproject_python/Client_Python/player/controller/HomepageController.py
from Client_Python.player.main.WaitingRoomMain import WaitingRoomMain
from Client_Python.player.main.LeaderboardsMain import LeaderboardsMain
from Client_Python.player.model import HomepageModel
from Client_Python.manager.PlayerCorbaManager import PlayerCorbaManager

import threading
import time
import sys

class HomepageController:
    def __init__(self):
        self.model = HomepageModel()
        self.corba_manager = PlayerCorbaManager.get_instance()

        # Start session validation timer
        self.session_timer_thread = None
        self.stop_session_check = threading.Event()
        self.session_expired = threading.Event()
        self.start_session_validation_timer()


    def handle_choice(self, choice, username):
        if choice == 1:
            print("-----Waiting Room-----\n")
            waiting_room = WaitingRoomMain(username) # Open Waiting Room
            waiting_room.wait_until_done()
            print("past waiting room main callable")#debug

        elif choice == 2:
            print("Viewing leaderboards...\n")
            LeaderboardsMain()  # Open Leaderboards
        elif choice == 3:
            print("Logging out...")

            self.stop_session_check.set()
            try:
                session_id = self.corba_manager.get_session_id()
                self.corba_manager.get_login_service().logoutPlayer(session_id)
                print(f"[INFO] Logged out session: {session_id}")
            except Exception as e:
                print(f"[ERROR] Failed to logout from server: {e}")

            from Client_Python.player.main.LoginMain import LoginMain
            LoginMain()
            raise SystemExit()
        else:
            print("Invalid choice, please select a valid option.")

    def start_session_validation_timer(self):
        def check_session():
            while not self.stop_session_check.is_set():
                try:
                    if not self.model.is_session_valid(self.corba_manager):
                        print("\n[WARNING] Your session has expired or you have been logged out elsewhere.\n")

                        session_id = self.corba_manager.get_session_id()
                        try:
                            self.corba_manager.get_login_service().logoutPlayer(session_id)
                        except Exception as logout_err:
                            print(f"[ERROR] Failed to logout from server: {logout_err}")

                        self.stop_session_check.set()
                        self.session_expired.set()  # <== notify main thread

                        return
                except Exception as e:
                    print(f"[ERROR] Session validation failed: {e}")
                time.sleep(5)

        self.session_timer_thread = threading.Thread(target=check_session, daemon=True)
        self.session_timer_thread.start()

File: 2025-9322-team04_finproject_java/src/main/java/Client_Java/player/controller/HomepageController.java
package Client_Java.player.controller;

import Client_Java.player.model.*;
import Client_Java.player.view.*;
import Client_Java.player.PlayerCorbaManager;
import compilations.*;

import javax.swing.*;

public class HomepageController {
    private final HomepageModel model;
    private final HomepageView view;
    private final PlayerCorbaManager corbaManager;

    public HomepageController(HomepageModel model, HomepageView view) {
        this.model = model;
        this.view = view;

        // Initialize CORBA manager with proper parameters
//        String[] corbaArgs = {"-ORBInitialPort", "1050", "-ORBInitialHost", "localhost"};
        this.corbaManager = PlayerCorbaManager.getInstance();

        setupActions();
        startSessionValidationTimer();
    }
    private void setupActions() {
        view.getStartGameBtn().addActionListener(e -> {
            view.getFrame().setVisible(false); // Hide instead of dispose
            openWaitingRoom();
        });

        view.getViewLeaderboardBtn().addActionListener(e -> {
            view.getFrame().setVisible(false); // Hide instead of dispose
            LeaderboardsController controller = new LeaderboardsController(this::returnToHomepage);
            controller.showLeaderboard();
        });

        view.getLogoutBtn().addActionListener(e -> logout());

        view.getMinimizeBtn().addActionListener(e -> view.getFrame().setState(JFrame.ICONIFIED));
        view.getCloseBtn().addActionListener(e -> System.exit(0));
        view.getMaximizeBtn().addActionListener(e -> toggleMaximize());
    }

    private void returnToHomepage() {
        SwingUtilities.invokeLater(() -> {
            view.getFrame().setVisible(true);
        });
    }

    private void openWaitingRoom() {
        SwingUtilities.invokeLater(() -> {
            try {
                // Get all necessary CORBA services
                GameService gameService = corbaManager.getGameService();
                WordService wordService = corbaManager.getWordService();
                AdminService adminService = corbaManager.getAdminService();

                String username = model.getUsername();
                if (username == null || username.isEmpty()) {
                    JOptionPane.showMessageDialog(null,
                            "Username is not set. Please login first.",
                            "Error",
                            JOptionPane.ERROR_MESSAGE);
                    return;
                }

                int waitTime = adminService.getJoinTimeout() > 0 ?
                        adminService.getJoinTimeout() : 10;

                WaitingRoomModel wrModel = new WaitingRoomModel(username, waitTime);
                WaitingRoomView wrView = new WaitingRoomView();

                // Pass all necessary services to WaitingRoomController
                new WaitingRoomController(wrModel, wrView, () -> {
                    view.getFrame().setVisible(true);
                }, gameService, wordService);

            } catch (Exception e) {
                e.printStackTrace();
                JOptionPane.showMessageDialog(null,
                        "Failed to connect to GameService.\nDetails: " + e.getMessage(),
                        "Connection Error",
                        JOptionPane.ERROR_MESSAGE);
                view.getFrame().setVisible(true);
            }
        });
    }

    public void openGameProper(String wordToPlay, int currentRound,
                               WordService wordService, GameService gameService,
                               String username, String gameId) {
        SwingUtilities.invokeLater(() -> {
            GameProperModel model = new GameProperModel(wordToPlay);
            GameProperView view = new GameProperView(wordToPlay);
            new GameProperController(model, view, corbaManager, username, currentRound, gameId);
        });
    }

    private void logout() {
        view.getFrame().dispose();

        String sessionId = corbaManager.getInstance().getSessionId();
        corbaManager.getInstance().getLoginService().logoutPlayer(sessionId);

        PlayerLoginView view = new PlayerLoginView();
        PlayerLoginModel model = new PlayerLoginModel();
        new PlayerLoginController(model, view);
    }

    private void toggleMaximize() {
        JFrame frame = view.getFrame();
        int state = frame.getExtendedState();
        boolean isCurrentlyMaximized = (state & JFrame.MAXIMIZED_BOTH) == JFrame.MAXIMIZED_BOTH;

        if (isCurrentlyMaximized) {
            frame.setExtendedState(JFrame.NORMAL);
            frame.setSize(1280, 720);
            frame.setLocationRelativeTo(null);
        } else {
            frame.setExtendedState(JFrame.MAXIMIZED_BOTH);
        }

        view.updateMaximizeButton(!isCurrentlyMaximized);
    }

    private void startSessionValidationTimer() {
        Timer sessionCheckTimer = new Timer(5000, e -> {
            try {
                if (!model.isSessionValid()) {
                    String sessionId = corbaManager.getInstance().getSessionId();
                    corbaManager.getInstance().getLoginService().logoutPlayer(sessionId); // Invalidate session

                    JOptionPane.showMessageDialog(view.getFrame(),
                            "Your session has expired or you have been logged out elsewhere.",
                            "Session Expired",
                            JOptionPane.WARNING_MESSAGE);

                    view.getFrame().dispose(); // Close current view
                    PlayerLoginView view = new PlayerLoginView();
                    PlayerLoginModel model = new PlayerLoginModel();
                    new PlayerLoginController(model, view);
                }
            } catch (Exception ex) {
                System.err.println("Error checking session validity: " + ex.getMessage());
            }
        });
        sessionCheckTimer.start();
    }

}